management-api-agent-hcd-1.2.x/src/main/java/org/apache/cassandra/transport/UnixSocketServerHcd.java

```diff
@@ -97,7 +97,34 @@ protected void channelRead0(ChannelHandlerContext ctx, Message.Request request)
         // logger.info("Executing {} {} {}", request, connection.getVersion(),
         // request.getStreamId());
 
-        Message.Response r = request.execute(qstate, queryStartNanoTime);
+        // In Cassandra 4.1.6, CASSANDRA-19534 changed the Message.Request.execute method signature
+        // to take a new Dispatcher.RequestTime object instead of a primitive long for the Query
+        // start time. We'll need to introduce reflection here to create the correct Objects and
+        // make the correct calls based on which version of 4.1.x we are. For Cassandra 5.0, this
+        // patch was ported between 5.0-rc1 and 5.0-rc2.
+        Message.Response r = null;
+        try {
+          // First see if we have the Dispatcher.RequestTime class. If so, assume we are 4.1.6+
+          Class dispatcherRequestTime =
+              Class.forName("org.apache.cassandra.transport.Dispatcher$RequestTime");
+          // we are 4.1.6+, get Dispatcher.RequestTime.forImmediateExecution()
+          Method forImmediateExecution =
+              dispatcherRequestTime.getDeclaredMethod("forImmediateExecution");
+          Method requestExecute =
+              Message.Request.class.getDeclaredMethod(
+                  "execute", QueryState.class, dispatcherRequestTime);
+          r =
+              (Message.Response)
+                  requestExecute.invoke(request, qstate, forImmediateExecution.invoke(null));
+
+        } catch (ClassNotFoundException cfne) {
+          logger.debug(
```

Comment:
I guess we will never need this part of the code in HCD? Or in 5.0.

Reply:
Probably not, unless an HCD 1.2 image is built with a Converged Core 5 version that has specific commits that make the Dispatcher method signatures match this stuff here.
Honestly, this reflection stuff is not great, but the only alternative is to duplicate the code out into more agent submodules for specific point releases of the C*/DSE/HCD server code. They are pretty good at not changing the API within a <major>.<minor> release, but this isn't exactly public APIs we are using. But I digress....

Reply:
To add to this, the main reason I'm leaving this code in place is in the case HCD needs to release a new 1.2 version and my patch for it to upgrade the Converged Core version does not get merged before the release. Then we would be in a situation where the Agent is using a Dispatcher signature that doesn't exist in the HCD build.